vault/token_store.go

```diff
@@ -1103,100 +1111,131 @@ func (ts *TokenStore) lookupBySaltedAccessor(saltedAccessor string) (accessorEnt
 // handleTidy handles the cleaning up of leaked accessor storage entries and
 // cleaning up of leases that are associated to tokens that are expired.
 func (ts *TokenStore) handleTidy(req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
-	// List out all the accessors
-	saltedAccessorList, err := ts.view.List(accessorPrefix)
-	if err != nil {
-		return nil, fmt.Errorf("failed to fetch accessor entries: %v", err)
-	}
-
 	var tidyErrors *multierror.Error
 
-	// First, clean up secondary index entries that are no longer valid
-	parentList, err := ts.view.List(parentPrefix)
-	if err != nil {
-		return nil, fmt.Errorf("failed to fetch secondary index entries: %v", err)
-	}
-
-	// Scan through the secondary index entries; if there is an entry
-	// with the token's salt ID at the end, remove it
-	for _, parent := range parentList {
-		children, err := ts.view.List(parentPrefix + parent + "/")
+	if atomic.CompareAndSwapInt64(&ts.tidyLock, 0, 1) {
```

Comment:
Actually I was going to ask you to move that check in the expiration manager to the top of that function but then decided to leave it alone :-)